cdc/sink/codec/avro.go

```diff
@@ -160,30 +152,86 @@ func (a *AvroEventBatchEncoder) Size() int {
 	return sum
 }
 
-func avroEncode(table *model.TableName, manager *AvroSchemaManager, tableVersion uint64, cols []*model.Column, colInfos []rowcodec.ColInfo, tz *time.Location) (*avroEncodeResult, error) {
+func (a *AvroEventBatchEncoder) avroEncode(
+	ctx context.Context,
+	e *model.RowChangedEvent,
+	topic string,
+	key bool,
+) (*avroEncodeResult, error) {
+	var cols []*model.Column
+	var colInfos []rowcodec.ColInfo
+	var enableTiDBExtension bool
+	var schemaManager *AvroSchemaManager
+	var operation string
+	if key {
+		cols, colInfos = e.HandleKeyColInfos()
+		enableTiDBExtension = false
+		schemaManager = a.keySchemaManager
+	} else {
+		cols = e.Columns
+		colInfos = e.ColInfos
+		enableTiDBExtension = a.enableTiDBExtension
+		schemaManager = a.valueSchemaManager
+		if e.IsInsert() {
+			operation = "c"
+		} else if e.IsUpdate() {
+			operation = "u"
+		} else {
+			log.Error("unknown operation", zap.Reflect("rowChangedEvent", e))
+			return nil, cerror.ErrAvroEncodeFailed.GenWithStack("unknown operation")
+		}
+	}
+
+	namespace := getAvroNamespace(a.namespace, e.Table)
+
 	schemaGen := func() (string, error) {
-		schema, err := ColumnInfoToAvroSchema(table.Table, cols)
+		schema, err := rowToAvroSchema(
+			namespace,
+			e.Table.Table,
+			cols,
+			colInfos,
+			enableTiDBExtension,
+			a.decimalHandlingMode,
+			a.bigintUnsignedHandlingMode,
+		)
 		if err != nil {
 			return "", errors.Annotate(err, "AvroEventBatchEncoder: generating schema failed")
 		}
```

Comment:
the error here, can only be `cerror.WrapError(cerror.ErrAvroMarshalFailed, err)`.

just use `errors.Trace(err)` should be ok.